### PR TITLE
CFE-4071: Added --masterfiles option for cfbs init

### PR DESCRIPTION
**Pull request by larsewi**

### CODE REVIEW

File: cfbs/commands.py [hunk 1]
```diff
@@ -228,24 +229,48 @@ def init_command(index=None, non_interactive=False) -> int:
...
+        to_add = "masterfiles@%s" % masterfiles
+    elif masterfiles != "no":
+        log.debug("--masterfiles=%s appears to be a branch" % masterfiles)
+        branch = masterfiles
+        to_add = "masterfiles"
 
+    ret = 0
     if to_add:
-        return add_command([to_add])
+        ret |= add_command([to_add])
```
Comment (vpodzime) [inline cfbs/commands.py:258]: Are you sure you want to do binary or on integers here?

File: cfbs/commands.py [hunk 2]
```diff
@@ -228,24 +229,48 @@ def init_command(index=None, non_interactive=False) -> int:
...
+                "Specify policy set to use instead (empty to skip)",
+                default="",
+            )
+    elif re.match(r"[0-9]+(\.[0-9]+){2}(\-[0-9]+)?", masterfiles):
+        log.debug("--masterfiles=%s appears to be a version number" % masterfiles)
+        to_add = "masterfiles@%s" % masterfiles
+    elif masterfiles != "no":
+        log.debug("--masterfiles=%s appears to be a branch" % masterfiles)
+        branch = masterfiles
+        to_add = "masterfiles"
```
Comment (vpodzime) [inline cfbs/commands.py:254]: I think there should be a commit here describing the magic happening in this case (and why).

**Review by vpodzime — COMMENTED** [review]
We should make sure that it's clear that the current `HEAD` of a branch is used, not the branch itself.

File: cfbs/commands.py (hunk 2 above)
Comment (larsewi) [reply cfbs/commands.py:259]: I'm guessing you meant comment, not commit?

File: cfbs/commands.py [hunk 3]
```diff
@@ -228,22 +229,52 @@ def init_command(index=None, non_interactive=False) -> int:
...
+
+    if branch is not None:
+        config = CFBSConfig.get_instance()
+        module = config.get_module_from_build("masterfiles")
+        remote = module["repo"]
+        commit = ls_remote(remote, branch)
+        if commit is None:
+            user_error("Failed to add masterfiles from branch %s" % branch)
+        log.debug("Current commit for masterfiles branch %s is %s" % (branch, commit))
+        module["url"] = remote
```
Comment (olehermanse) [inline cfbs/commands.py:274]: Note / Todo: We should store the branch name here to enable `cfbs update`. But we can do this in a separate PR.

File: cfbs/main.py [hunk 4]
```diff
@@ -45,6 +45,11 @@ def main() -> int:
         print("")
         user_error("No command given")
 
+    if args.masterfiles and args.command != "init":
+        user_error(
+            "The option --masterfiles is only for cfbs init, not cfbs %s" % args.command
```
Comment (olehermanse) [inline cfbs/main.py:50]: I think it would read a bit better with quotes:

```suggestion
            "The option --masterfiles is only for 'cfbs init', not 'cfbs %s'" % args.command
```

File: cfbs/cfbs_config.py [hunk 5]
```diff
@@ -81,7 +81,7 @@ def save(self):
             f.write(data)
 
     def longest_module_name(self) -> int:
-        return max((len(m["name"]) for m in self["build"]))
+        return max((len(m["name"]) for m in self["build"])) if self["build"] else 0
```
Comment (olehermanse) [inline cfbs/cfbs_config.py:84]: 👍

File: tests/shell/026_init_no_masterfiles.sh [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+set -x
+cd tests/
+mkdir -p ./tmp/
+cd ./tmp/
+touch cfbs.json && rm cfbs.json
+rm -rf .git
+
+cfbs --non-interactive init --masterfiles=no
+!( grep '"name": "masterfiles"' cfbs.json )
+cfbs build
```
Comment (olehermanse) [inline tests/shell/026_init_no_masterfiles.sh:11]: TODO: We should probably add a warning or error when building an empty project.

File: tests/shell/028_init_masterfiles_version_3.18.2.sh [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+set -x
+cd tests/
+mkdir -p ./tmp/
+cd ./tmp/
+touch cfbs.json && rm cfbs.json
+rm -rf .git
+
+cfbs --non-interactive init --masterfiles=3.18.2
+grep '"name": "masterfiles"' cfbs.json
+grep '"version": "3.18.2"' cfbs.json
```
Comment (olehermanse) [inline tests/shell/028_init_masterfiles_version_3.18.2.sh:11]: Since 3.18.2 isn't gonna move, I think it would be nice to grep for the commit SHA as well here.

File: tests/shell/029_init_masterfiles_version_3.18.1-1.sh [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+cd tests/
+mkdir -p ./tmp/
+cd ./tmp/
+touch cfbs.json && rm cfbs.json
+rm -rf .git
+
+cfbs --non-interactive init --masterfiles=3.18.1-1
+grep '"name": "masterfiles"' cfbs.json
+grep '"version": "3.18.1-1"' cfbs.json
+cfbs build
```
Comment (olehermanse) [inline tests/shell/029_init_masterfiles_version_3.18.1-1.sh:12]:
```suggestion
grep '"version": "3.18.1-1"' cfbs.json
cfbs build

cfbs update
grep '"version"' cfbs.json
grep '"version": "3.18.1-1"' cfbs.json
```

File: cfbs/commands.py (hunk 3 above)
Comment (larsewi) [reply cfbs/commands.py:274]: [CFE-4089](https://tracker.mender.io/browse/CFE-4089)

File: tests/shell/026_init_no_masterfiles.sh (hunk 6 above)
Comment (larsewi) [reply tests/shell/026_init_no_masterfiles.sh:11]: [CFE-4090](https://tracker.mender.io/browse/CFE-4090)

File: cfbs/cfbs_config.py (hunk 5 above)
Comment (larsewi) [reply cfbs/cfbs_config.py:84]: :+1:

**Review by olehermanse — COMMENTED** [review]
Looks good in general! I added some smaller non-critical comments. There is a conflict now, but luckily, it's very easy to fix :)

**Review by olehermanse — APPROVED** [review]
Conflict fixed, ready to merge (Lars can submit follow-ups).

File: tests/shell/028_init_masterfiles_version_3.18.2.sh (hunk 7 above)
Comment (larsewi) [reply tests/shell/028_init_masterfiles_version_3.18.2.sh:11]: https://github.com/cfengine/cfbs/pull/138

File: tests/shell/029_init_masterfiles_version_3.18.1-1.sh (hunk 8 above)
Comment (larsewi) [reply tests/shell/029_init_masterfiles_version_3.18.1-1.sh:12]: https://github.com/cfengine/cfbs/pull/138
